protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,233 @@
+For every height $h$:
+
+- _agreement_: no two correct processes decide differently.
+- _validity_: function $valid(v, s)$, when $v$ is the decided block, always returns _true_.
+- _termination_: all correct processes eventually decide.
+
+The _validity_ property refers to $valid(v, bc_{h-1})$, which is defined as a (mathematical) function
+whose output depends exclusively on its inputs: the value proposed $v$ (i.e., a block),
+and the state of the blockchain after applying the block decided at height $h-1$,
+denoted $bc_{h-1}$.
+
+The consensus algorithm implemented in CometBFT (Tendermint) fulfills these properties.
+
+### ABCI interface
+
+In the version of ABCI (`v0.17.0`) that existed before ABCI 1.0 and 2.0 (a.k.a. ABCI++),
+the implementation of function $valid(v, bc_{h-1})$ was totally internal to CometBFT.
+Technically, the application's part of $bc_{h-1}$ was not considered by function $valid()$.
+Thus, the application had no direct say on the validity of a block,
+although it could (and still can) indirectly influence the contents of blocks via the (best-effort) ABCI call `CheckTx`.
+
+With the evolution of ABCI to ABCI 1.0 and 2.0, CometBFT's implementation of
+function $valid(v, bc_{h-1})$ now has two components:
+
+- the validity checks performed directly by CometBFT on blocks
+  (block format, hashes, etc; the same as in ABCI `v0.17.0`)
+- the validity checks that now the application can perform as part of `ProcessProposal`;
+  i.e., `ProcessProposal` is now part of $valid(v, bc_{h-1})$
+
+With the new structure of the implementation of function $valid(v, bc_{h-1})$:
+
+- consensus _agreement_ is not affected and all processes are still required to agree on the same value
+- the consensus _validity_ property is not affected since we are changing the
+  internals of function $valid(v, bc_{h-1})$; consensus _validity_ just requires this function to be true
+
+However, the new structure of the implementation of function $valid(v, bc_{h-1})$
+may affect _termination_ of consensus, as some implementations of `ProcessProposal` might reject values
+that CometBFT's internal validity checks would otherwise accept.
+
+This document focuses on how consensus _termination_ is affected
+by the new structure of function $valid(v, bc_{h-1})$,
+in particular, the different implementations of `ProcessProposal`.
+
+### ABCI 1.0 (and 2.0) Specification
+
+The [ABCI 1.0 specification][abci-spec] imposes a set of new requirements on the application
+so that its implementation of `PrepareProposal` and `ProcessProposal` does not compromise _termination_ of consensus,
+given the current CometBFT consensus algorithm
+(called Tendermint, and described in the [arXiv paper][arxiv]).
+In contrast to $valid(v, bc_{h-1})$, which is defined as a mathematical function used for consensus's formal specification,
+`PrepareProposal` and `ProcessProposal` are understood as implemented functions in CometBFT.
```

Comment:
I believe this has to be reviewed, in the light of the more recent changes. valid now is split into two components. Also, nowhere so far it has been said that ProcessProposal does not have to be a pure function.

Reply:
What would you suggest as text?

Reply:
"implemented function" means function in its software (rather than mathematical) meaning. I'll try to reword